pkg/traversal/traversal.go

```diff
@@ -1,59 +1,33 @@
 package traversal
 
 import (
+	"io/fs"
 	"path/filepath"
-	"sort"
 	"sync"
 
 	"k8s.io/klog/v2"
-
-	"github.com/openshift/must-gather-clean/pkg/input"
-	"github.com/openshift/must-gather-clean/pkg/obfuscator"
-	"github.com/openshift/must-gather-clean/pkg/omitter"
-	"github.com/openshift/must-gather-clean/pkg/output"
 )
 
 type FileWalker struct {
-	obfuscators  []obfuscator.Obfuscator
-	fileOmitters []omitter.FileOmitter
-	k8sOmitters  []omitter.KubernetesResourceOmitter
-	reader       input.Inputter
-	workerCount  int
-	workers      []*worker
-	writer       output.Outputter
-}
-
-// NewFileWalker returns a FileWalker. It ensures that the reader directory exists and is readable.
-func NewFileWalker(reader input.Inputter,
-	writer output.Outputter,
-	obfuscators []obfuscator.Obfuscator,
-	fileOmitters []omitter.FileOmitter,
-	k8sOmitters []omitter.KubernetesResourceOmitter,
-	workerCount int) (*FileWalker, error) {
-	return &FileWalker{
-		obfuscators:  obfuscators,
-		fileOmitters: fileOmitters,
-		k8sOmitters:  k8sOmitters,
-		reader:       reader,
-		workerCount:  workerCount,
-		writer:       writer,
-	}, nil
+	inputPath     string
+	workerCount   int
+	workers       []QueueProcessor
+	workerFactory func(int) QueueProcessor
 }
 
-// Traverse should be called to start processing the reader directory.
+// Traverse should be called to start processing the must-gather directory. This method will exist the CLI if an error is encountered.
 func (w *FileWalker) Traverse() {
 	wg := sync.WaitGroup{}
 	errorCh := make(chan error, w.workerCount)
-	queue := make(chan workerFile, w.workerCount)
-	w.workers = make([]*worker, w.workerCount)
+	queue := make(chan WorkerInput, w.workerCount)
+	w.workers = make([]QueueProcessor, w.workerCount)
 	for i := 0; i < w.workerCount; i++ {
-		wk := newWorker(i+1, w.obfuscators, w.fileOmitters, w.k8sOmitters, queue, w.writer, errorCh)
-		w.workers[i] = wk
+		w.workers[i] = w.workerFactory(i + 1)
 		wg.Add(1)
-		go func() {
-			wk.run()
+		go func(i int, queue chan WorkerInput, errorCh chan error) {
```

Comment:
Does `WorkerInput` have to be a public type? Also what's the reason for passing them in here in `ProcessQueue()` instead of the new function?

Reply:
made workerInput private, thanks. 

> Also what's the reason for passing them in here in ProcessQueue() instead of the new function?

it's to decouple the worker logic from business logic. Otherwise the worker factory would need to know that there is a work channel and some error channel. The worker is created in the cli.go now, there's no notion of channels - will make it clearer by completely pulling out all business logic from the traversal package. Then it's purpose is to only handle the channels and take the files from the traversal.